<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Pandas-Assignment" data-toc-modified-id="Pandas-Assignment-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Pandas Assignment</a></span><ul class="toc-item"><li><span><a href="#How-many-rows-of-data-are-in-the-DataFrame?" data-toc-modified-id="How-many-rows-of-data-are-in-the-DataFrame?-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>How many rows of data are in the DataFrame?</a></span></li><li><span><a href="#What-are-the-names-and-data-types-of-the-columns?" data-toc-modified-id="What-are-the-names-and-data-types-of-the-columns?-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>What are the names and data types of the columns?</a></span></li><li><span><a href="#Do-any-of-the-columns-contain-null-values?" data-toc-modified-id="Do-any-of-the-columns-contain-null-values?-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Do any of the columns contain null values?</a></span></li><li><span><a href="#Find-all-successful-documentary-projects-and-sort-them-by-the-amount-pledged.-Print-the-top-10-highest-pledges." data-toc-modified-id="Find-all-successful-documentary-projects-and-sort-them-by-the-amount-pledged.-Print-the-top-10-highest-pledges.-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.</a></span></li><li><span><a href="#Create-a-new-column-named-average_per_backer-and-set-its-value-to-the-total-amount-pledged-/-number-of-backers." data-toc-modified-id="Create-a-new-column-named-average_per_backer-and-set-its-value-to-the-total-amount-pledged-/-number-of-backers.-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Create a new column named average_per_backer and set its value to the total amount pledged / number of backers.</a></span></li><li><span><a href="#What-happened-to-the-rows-with-0-backers?-How-can-this-be-dealt-with?" data-toc-modified-id="What-happened-to-the-rows-with-0-backers?-How-can-this-be-dealt-with?-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>What happened to the rows with 0 backers? How can this be dealt with?</a></span></li><li><span><a href="#Drop-all-rows-with-0-backers-then-repeat-the-previous-exercise." data-toc-modified-id="Drop-all-rows-with-0-backers-then-repeat-the-previous-exercise.-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Drop all rows with 0 backers then repeat the previous exercise.</a></span></li><li><span><a href="#Create-a-DataFrame-whose-index-is-the-first-10-letters-of-the-alphabet-and-contains-two-more-columns-with-first-10-prime-numbers-and-the-first-10-fibonacci-numbers." data-toc-modified-id="Create-a-DataFrame-whose-index-is-the-first-10-letters-of-the-alphabet-and-contains-two-more-columns-with-first-10-prime-numbers-and-the-first-10-fibonacci-numbers.-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.</a></span></li><li><span><a href="#Merge-two-DataFrames-(flights,-airlines)-into-a-single-DataFrame:-nycflights13." data-toc-modified-id="Merge-two-DataFrames-(flights,-airlines)-into-a-single-DataFrame:-nycflights13.-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Merge two DataFrames (flights, airlines) into a single DataFrame: nycflights13.</a></span></li><li><span><a href="#Are-there-any-missing-values-in-this-DataFrame?-If-so,-drop-them." data-toc-modified-id="Are-there-any-missing-values-in-this-DataFrame?-If-so,-drop-them.-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>Are there any missing values in this DataFrame? If so, drop them.</a></span></li><li><span><a href="#Merge-these-two-DataFrames-into-a-single-DataFrame-called-census." data-toc-modified-id="Merge-these-two-DataFrames-into-a-single-DataFrame-called-census.-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>Merge these two DataFrames into a single DataFrame called census.</a></span></li><li><span><a href="#Re-shape-census-so-that-one-column-contains-all-population-measures,-and-another-the-year-attributes." data-toc-modified-id="Re-shape-census-so-that-one-column-contains-all-population-measures,-and-another-the-year-attributes.-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>Re-shape census so that one column contains all population measures, and another the year attributes.</a></span></li><li><span><a href="#Group-the-data-by-year-and-summarize-it." data-toc-modified-id="Group-the-data-by-year-and-summarize-it.-1.13"><span class="toc-item-num">1.13&nbsp;&nbsp;</span>Group the data by year and summarize it.</a></span></li><li><span><a href="#Group-the-data-by-region,-division-and-year-and-summarize-it." data-toc-modified-id="Group-the-data-by-region,-division-and-year-and-summarize-it.-1.14"><span class="toc-item-num">1.14&nbsp;&nbsp;</span>Group the data by region, division and year and summarize it.</a></span></li><li><span><a href="#Create-a-crosstab-between-state-and-category-showing-amount-pledged-as-the-value." data-toc-modified-id="Create-a-crosstab-between-state-and-category-showing-amount-pledged-as-the-value.-1.15"><span class="toc-item-num">1.15&nbsp;&nbsp;</span>Create a crosstab between state and category showing amount pledged as the value.</a></span></li></ul></li></ul></div>

To:&nbsp;&nbsp;&nbsp;&nbsp; Magnimind

From: Matt Curcio, matt.curcio.ri@gmail.com

Date: 2022-10-01

Re:&nbsp;&nbsp;&nbsp; Intro. to Python

In [1]:
import pandas as pd
import numpy as np

## Pandas Assignment

Please refer to the following link for this practice:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv

### How many rows of data are in the DataFrame?


In [2]:
!pwd
filename = 'ksprojects.csv'

/home/mcc/Dropbox/9-Magni-mind/8_Pandas_Assignment


In [3]:
!ls

'2022-10-19 Magnimind Pandas Assignment.ipynb'	 ksprojects.csv
 airlines.csv					 panda_questions.txt
 census-divisions.csv				 state-populations.csv
 flights.csv					'Testing argsort.ipynb'


In [4]:
filename = 'ksprojects.csv'

ksp = pd.read_csv(filename)
ksp[0:2]

id                                               name     category  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...  Documentary   
1  1100844465                                       daily digest  Documentary   

  main_category             deadline   goal             launched       state  \
0  Film & Video  2009-05-20 21:50:00  300.0  2009-04-30 22:10:30      failed   
1  Film & Video  2009-06-01 02:20:00  700.0  2009-05-04 21:14:28  successful   

   backers country  usd_pledged currency  pledged  
0        4      US         40.0      USD     40.0  
1       14      US        700.0      USD    700.0

In [5]:
print(f'\n"ksprojects.csv" has {ksp.shape[0]} rows.\n')


"ksprojects.csv" has 65418 rows.



### What are the names and data types of the columns?

In [6]:
ksp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65418 entries, 0 to 65417
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             65418 non-null  int64  
 1   name           65418 non-null  object 
 2   category       65418 non-null  object 
 3   main_category  65418 non-null  object 
 4   deadline       65418 non-null  object 
 5   goal           65418 non-null  float64
 6   launched       65418 non-null  object 
 7   state          65418 non-null  object 
 8   backers        65418 non-null  int64  
 9   country        65418 non-null  object 
 10  usd_pledged    64673 non-null  float64
 11  currency       65418 non-null  object 
 12  pledged        65418 non-null  float64
dtypes: float64(3), int64(2), object(8)
memory usage: 6.5+ MB


### Do any of the columns contain null values?

In [7]:
ksp.isnull().sum()

id                 0
name               0
category           0
main_category      0
deadline           0
goal               0
launched           0
state              0
backers            0
country            0
usd_pledged      745
currency           0
pledged            0
dtype: int64

- **It appears that column 'usd_pledged' has 745 null values.**

### Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.

In [8]:
# The list of unique categories is very long since each film 
# can be a member of many categories.

#ksp['category'].unique()

In [9]:
# the number of film categories is

cats = len(ksp['category'].unique())

print(f'\nThe number of film categories is {cats}.\n')


The number of film categories is 158.



In [10]:
# The number of states (i.e. successful, ..., failed) is 6.

ksp['state'].unique()

array(['failed', 'successful', 'canceled', 'suspended', 'undefined',
       'live'], dtype=object)

In [11]:
# The number is films that are documentaries.

solo_docs = ksp[ksp['category'] == 'Documentary']

print(f'\nThe number of "Documentary" films is {solo_docs.shape[0]}.\n')


The number of "Documentary" films is 2969.



In [12]:
successful_docs = solo_docs[solo_docs['state'] == 'successful']

print(f'\nThe number of "Successful & Documentary" films is {successful_docs.shape[0]}.\n')



The number of "Successful & Documentary" films is 1087.



In [13]:
type(successful_docs)

pandas.core.frame.DataFrame

In [14]:
successful_docs[0:2]

id                                     name     category  \
1  1100844465                             daily digest  Documentary   
9  1125367305  Meeting Andrei Tarkovsky - Film Website  Documentary   

  main_category             deadline    goal             launched       state  \
1  Film & Video  2009-06-01 02:20:00   700.0  2009-05-04 21:14:28  successful   
9  Film & Video  2009-06-21 02:31:00  1250.0  2009-05-30 06:27:42  successful   

   backers country  usd_pledged currency  pledged  
1       14      US        700.0      USD    700.0  
9       60      US       2721.0      USD   2721.0

In [15]:
#Ex df.sort_values(by=['Brand'], inplace=True, ascending=False)

sorted = successful_docs.sort_values(by=['pledged'], inplace=False, ascending=False)

sorted[0:10]

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
12127  1326363492                                   The Culture High   
61248  1203470139                         Curious George Documentary   
28003  1205913248                                     ILEANA'S SMILE   
27402  1135999291                             Rewind To Fast-Forward   
16243  1061853312                               Linsanity: the Movie   
27268  1228065368                  El Voto Hispano | The Documentary   
43820  1194449879  PlantPure Nation, a Documentary with Dr. T. Co...   
22237   122290911                          The Killing of Tony Blair   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
61248  Documentary  Film & Video  2016-08-25 00:00:00  175000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
27402  Documentary  Film & Video  2014-05-01 07:59:00  137000.0   
16243  Documentary  Film & Video  2013-02-16 00:00:00  117000.0   
27268  Documentary  Film & Video  2014-04-27 08:15:23  158000.0   
43820  Documentary  Film & Video  2015-05-14 17:41:41  150000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US     384376.0      USD   
24166  2013-10-30 21:06:34  successful     4245      US     313158.0      USD   
12127  2012-06-22 01:40:20  successful     3448      US     240022.0      USD   
61248  2016-07-26 18:01:41  successful     1483      US       2028.0      USD   
28003  2014-04-21 09:05:00  successful     1173      US     177421.0      USD   
27402  2014-03-30 13:35:31  successful     4395      US     176746.0      USD   
16243  2013-01-15 19:46:34  successful     2102      US     167917.0      USD   
27268  2014-03-18 07:15:23  successful      461      US     165401.0      USD   
43820  2015-03-15 16:41:41  successful     1099      US     163930.0      USD   
22237  2013-08-28 18:17:14  successful     4459      GB     255150.0      GBP   

        pledged  
11731  384376.0  
24166  313158.0  
12127  240022.0  
61248  186010.0  
28003  177421.0  
27402  176746.0  
16243  167917.0  
27268  165401.0  
43820  163930.0  
22237  163891.0

### Create a new column named average_per_backer and set its value to the total amount pledged / number of backers.

average_per_backer = total amount pledged / number of backers = **(usd_pledged + pledged)/backers**

In [16]:
sorted['avg_per_backer'] = sorted['pledged']/sorted['backers']

sorted[0:5]

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
12127  1326363492                                   The Culture High   
61248  1203470139                         Curious George Documentary   
28003  1205913248                                     ILEANA'S SMILE   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
61248  Documentary  Film & Video  2016-08-25 00:00:00  175000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US     384376.0      USD   
24166  2013-10-30 21:06:34  successful     4245      US     313158.0      USD   
12127  2012-06-22 01:40:20  successful     3448      US     240022.0      USD   
61248  2016-07-26 18:01:41  successful     1483      US       2028.0      USD   
28003  2014-04-21 09:05:00  successful     1173      US     177421.0      USD   

        pledged  avg_per_backer  
11731  384376.0       59.062077  
24166  313158.0       73.771025  
12127  240022.0       69.611949  
61248  186010.0      125.428186  
28003  177421.0      151.254049

### What happened to the rows with 0 backers? How can this be dealt with?

In [17]:
sorted['backers'].isna().value_counts()

#sorted[0:1]

False    1087
Name: backers, dtype: int64

- **In the sorted-successful-documentary films there are NO null values or NA.**

In [18]:
sorted.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1087 entries, 11731 to 31857
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              1087 non-null   int64  
 1   name            1087 non-null   object 
 2   category        1087 non-null   object 
 3   main_category   1087 non-null   object 
 4   deadline        1087 non-null   object 
 5   goal            1087 non-null   float64
 6   launched        1087 non-null   object 
 7   state           1087 non-null   object 
 8   backers         1087 non-null   int64  
 9   country         1087 non-null   object 
 10  usd_pledged     1087 non-null   float64
 11  currency        1087 non-null   object 
 12  pledged         1087 non-null   float64
 13  avg_per_backer  1087 non-null   float64
dtypes: float64(4), int64(2), object(8)
memory usage: 127.4+ KB


In [19]:
ksp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65418 entries, 0 to 65417
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             65418 non-null  int64  
 1   name           65418 non-null  object 
 2   category       65418 non-null  object 
 3   main_category  65418 non-null  object 
 4   deadline       65418 non-null  object 
 5   goal           65418 non-null  float64
 6   launched       65418 non-null  object 
 7   state          65418 non-null  object 
 8   backers        65418 non-null  int64  
 9   country        65418 non-null  object 
 10  usd_pledged    64673 non-null  float64
 11  currency       65418 non-null  object 
 12  pledged        65418 non-null  float64
dtypes: float64(3), int64(2), object(8)
memory usage: 6.5+ MB


- **NOTE: The column 'backers' does not have any null values. Therefore the possible division by zero should NOT be a problem.**

### Drop all rows with 0 backers then repeat the previous exercise.

**Hold it**, I found no missing values in the backers column in either the original dataset or the sorted-successful-documentary' sub-set.

### Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.

Produce:
1. Column list of first 10 letters
2.  " list of first 10 primes
3.  " list of first 10 fibonacci numbers

In [20]:
lst = [chr(x) for x in range(97, 107)]
primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
fibs = [1, 1, 2, 3, 5, 8, 13, 21, 34, 55]

df = pd.DataFrame({'primes':primes,'fibs':fibs},index=lst)

df

primes  fibs
a       2     1
b       3     1
c       5     2
d       7     3
e      11     5
f      13     8
g      17    13
h      19    21
i      23    34
j      29    55

### Merge two DataFrames (flights, airlines) into a single DataFrame: nycflights13.

Please use the following dataset for merging purposes:

- https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv
- https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv  
- **OR**
- flights = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv')
- airline = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv')

In [21]:
filename1 = 'flights.csv'
flights = pd.read_csv(filename1)

print(f'\nThe dimensions of "flights.csv" is {flights.shape}.')
flights[0:2]


The dimensions of "flights.csv" is (336776, 19).


year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416     5      29  2013-01-01 05:00:00

In [22]:
filename2 = 'airlines.csv'
airlines = pd.read_csv(filename2)

print(f'\nThe dimensions of "airlines.csv" is {airlines.shape}.')
airlines[0:2]


The dimensions of "airlines.csv" is (16, 2).


carrier                    name
0      9E       Endeavor Air Inc.
1      AA  American Airlines Inc.

In [23]:
flights.isnull().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

In [24]:
airlines.isnull().sum()

carrier    0
name       0
dtype: int64

In [25]:
nycflights13 = pd.merge(flights, airlines, on = "carrier", how = "outer")

nycflights13[0:2]

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   

   distance  hour  minute            time_hour                   name  
0      1400     5      15  2013-01-01 05:00:00  United Air Lines Inc.  
1      1416     5      29  2013-01-01 05:00:00  United Air Lines Inc.

### Are there any missing values in this DataFrame? If so, drop them.

In [26]:
nycflights13.isna().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
name                 0
dtype: int64

In [27]:
nycflights13_dropna = nycflights13.dropna() # Drop Rows with NA

In [28]:
nycflights13_dropna.isna().sum()

year              0
month             0
day               0
dep_time          0
sched_dep_time    0
dep_delay         0
arr_time          0
sched_arr_time    0
arr_delay         0
carrier           0
flight            0
tailnum           0
origin            0
dest              0
air_time          0
distance          0
hour              0
minute            0
time_hour         0
name              0
dtype: int64

### Merge these two DataFrames into a single DataFrame called census.

Please use the following dataset for merging purposes:

- https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv

- https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv

In [29]:
!ls

'2022-10-19 Magnimind Pandas Assignment.ipynb'	 ksprojects.csv
 airlines.csv					 panda_questions.txt
 census-divisions.csv				 state-populations.csv
 flights.csv					'Testing argsort.ipynb'


In [30]:
state_pop = pd.read_csv('state-populations.csv')

print(f'\nThe dimensions of "state-populations.csv" is {state_pop.shape}.')
state_pop[0:2]


The dimensions of "state-populations.csv" is (51, 8).


state     2010     2011     2012     2013     2014     2015     2016
0  Alabama  4785492  4799918  4815960  4829479  4843214  4853875  4863300
1   Alaska   714031   722713   731089   736879   736705   737709   741894

In [31]:
census_div = pd.read_csv('census-divisions.csv')

print(f'\nThe dimensions of "census-divisions.csv" is {census_div.shape}.')
census_div[0:2]


The dimensions of "census-divisions.csv" is (51, 3).


state     region     division
0  Connecticut  Northeast  New England
1        Maine  Northeast  New England

In [32]:
census = pd.merge(state_pop, census_div, on = "state", how = "outer")

print(f'\nThe dimensions of "census" dataframe is {census.shape}.')
census[0:2]


The dimensions of "census" dataframe is (51, 10).


state     2010     2011     2012     2013     2014     2015     2016  \
0  Alabama  4785492  4799918  4815960  4829479  4843214  4853875  4863300   
1   Alaska   714031   722713   731089   736879   736705   737709   741894   

  region            division  
0  South  East South Central  
1   West             Pacific

### Re-shape census so that one column contains all population measures, and another the year attributes.

In [33]:
shaped_census = census.melt(id_vars=['state','region','division'],var_name='year',value_name='population')
shaped_census.head()

state region            division  year  population
0     Alabama  South  East South Central  2010     4785492
1      Alaska   West             Pacific  2010      714031
2     Arizona   West            Mountain  2010     6408312
3    Arkansas  South  West South Central  2010     2921995
4  California   West             Pacific  2010    37332685

In [34]:
shaped_census.tail()

state   region            division  year  population
352       Virginia    South      South Atlantic  2016     8411808
353     Washington     West             Pacific  2016     7288000
354  West Virginia    South      South Atlantic  2016     1831102
355      Wisconsin  Midwest  East North Central  2016     5778708
356        Wyoming     West            Mountain  2016      585501

### Group the data by year and summarize it.

In [35]:
shaped_census.groupby('year').describe()

population                                                              \
          count          mean           std       min        25%        50%   
year                                                                          
2010       51.0  6.065651e+06  6.839536e+06  564513.0  1700530.5  4348662.0   
2011       51.0  6.111046e+06  6.907730e+06  567725.0  1713213.0  4369354.0   
2012       51.0  6.156831e+06  6.974942e+06  576765.0  1725818.0  4384799.0   
2013       51.0  6.200096e+06  7.038774e+06  582684.0  1732621.0  4400477.0   
2014       51.0  6.246342e+06  7.110811e+06  583642.0  1741023.0  4413057.0   
2015       51.0  6.292091e+06  7.181603e+06  586555.0  1746940.5  4424611.0   
2016       51.0  6.335834e+06  7.243849e+06  585501.0  1757121.0  4436974.0   

                             
            75%         max  
year                         
2010  6654375.0  37332685.0  
2011  6717221.5  37676861.0  
2012  6776617.0  38011074.0  
2013  6837396.0  38335203.0  
2014  6902053.5  38680810.0  
2015  6988927.5  38993940.0  
2016  7109535.5  39250017.0

### Group the data by region, division and year and summarize it.

In [36]:
shaped_census.groupby(['region','division','year']).describe()

population                            \
                                     count        mean           std   
region  division           year                                        
Midwest East North Central 2010        5.0   9288169.4  3.115322e+06   
                           2011        5.0   9301433.8  3.109807e+06   
                           2012        5.0   9314559.0  3.104953e+06   
                           2013        5.0   9332093.0  3.099620e+06   
                           2014        5.0   9346265.8  3.091076e+06   
...                                    ...         ...           ...   
West    Pacific            2012        5.0  10185665.0  1.574263e+07   
                           2013        5.0  10274464.0  1.587648e+07   
                           2014        5.0  10371286.2  1.601909e+07   
                           2015        5.0  10468346.0  1.614496e+07   
                           2016        5.0  10560386.6  1.624360e+07   

                                                                              \
                                       min        25%        50%         75%   
region  division           year                                                
Midwest East North Central 2010  5690263.0  6490528.0  9877495.0  11540983.0   
                           2011  5709640.0  6516480.0  9876213.0  11544824.0   
                           2012  5726177.0  6537743.0  9887238.0  11550839.0   
                           2013  5742854.0  6569102.0  9898982.0  11570022.0   
                           2014  5758377.0  6595233.0  9915767.0  11594408.0   
...                                    ...        ...        ...         ...   
West    Pacific            2012   731089.0  1391820.0  3899116.0   6895226.0   
                           2013   736879.0  1406481.0  3925751.0   6968006.0   
                           2014   736705.0  1416349.0  3968371.0   7054196.0   
                           2015   737709.0  1425157.0  4024634.0   7160290.0   
                           2016   741894.0  1428557.0  4093465.0   7288000.0   

                                             
                                        max  
region  division           year              
Midwest East North Central 2010  12841578.0  
                           2011  12860012.0  
                           2012  12870798.0  
                           2013  12879505.0  
                           2014  12867544.0  
...                                     ...  
West    Pacific            2012  38011074.0  
                           2013  38335203.0  
                           2014  38680810.0  
                           2015  38993940.0  
                           2016  39250017.0  

[63 rows x 8 columns]

### Create a crosstab between state and category showing amount pledged as the value.

Please refer to the following link for this practice:

- https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv

In [37]:
ksp = pd.read_csv('ksprojects.csv')
ksp.columns

Index(['id', 'name', 'category', 'main_category', 'deadline', 'goal',
       'launched', 'state', 'backers', 'country', 'usd_pledged', 'currency',
       'pledged'],
      dtype='object')

In [38]:
pd.crosstab(ksp.state, ksp.category, values=ksp.pledged, aggfunc='sum')


category    3D Printing   Academic  Accessories    Action   Animals  \
state                                                                 
canceled      106827.50    4364.00    124936.11   13342.0      0.00   
failed        371417.00   81427.74    320990.20   79978.1  15769.87   
live          309215.00      31.00     60904.00       0.0       NaN   
successful   8511414.01  354485.24   2569489.70  226333.9  75497.10   
suspended      10146.00     226.00      3822.00       NaN      0.00   
undefined      13757.00        NaN      2763.00       NaN       NaN   

category     Animation  Anthologies     Apparel       Apps  Architecture  ...  \
state                                                                     ...   
canceled      58910.00      4583.00   228836.00  174745.49      12311.00  ...   
failed       498129.57     20991.00   401102.55  553596.12     139656.99  ...   
live          98403.80      3339.00   236921.59   60101.00       2117.00  ...   
successful  2605326.54   1346909.64  6446055.31  755165.60     460293.08  ...   
suspended         0.00          NaN    56213.00   12989.00        131.00  ...   
undefined          NaN          NaN         NaN        NaN           NaN  ...   

category    Wearables  Weaving        Web   Webcomics   Webseries  \
state                                                               
canceled     239056.0    212.0   88658.00     5668.00    54679.39   
failed       718671.5   6686.0  382408.65    26085.00  1065147.12   
live         141310.0      NaN   57931.00    82120.00      185.00   
successful  7381979.5   6019.0  912191.70  1164701.39  3821840.97   
suspended      9344.0      0.0      50.00         NaN    14878.00   
undefined         NaN      NaN        NaN         NaN         NaN   

category    Woodworking  Workshops  World Music  Young Adult     Zines  
state                                                                   
canceled         153.01        NaN     17136.00       661.00   6940.00  
failed         60328.00     2360.0    120691.83     20920.20  19692.00  
live            1211.73        0.0      9580.00       575.00     80.00  
successful    214568.35    21885.0   1069703.20    145305.62  48668.53  
suspended         91.00        0.0        37.00          NaN       NaN  
undefined           NaN        NaN          NaN          NaN       NaN  

[6 rows x 158 columns]

pd.crosstab(index, columns, 
                values=None, rownames=None, 
                colnames=None, aggfunc=None, 
                margins=False, margins_name: str = 'All', 
                dropna: bool = True, normalize=False)